examples/vision/knowledge_distillation.py

```diff
@@ -0,0 +1,276 @@
+
+        # Return a dict of performance
+        results = {m.name: m.result() for m in self.metrics}
+        results.update({"student_loss": student_loss})
+        return results
+
+
+"""
+## Create student and teacher models
+Initialy, we create a teacher model and a smaller student model. Both models are
+convolutional neural networks and created using `Sequential()`, but could be any model
+created with the Keras API.
+"""
+
+# Create the teacher
+teacher = keras.Sequential(
+    [
+        keras.Input(shape=(28, 28, 1)),
+        layers.Conv2D(256, (3, 3), strides=(2, 2), padding="same"),
+        layers.LeakyReLU(alpha=0.2),
+        layers.MaxPooling2D(pool_size=(2, 2), strides=(1, 1), padding="same"),
+        layers.Conv2D(512, (3, 3), strides=(2, 2), padding="same"),
+        layers.Flatten(),
+        layers.Dense(10),
+    ],
+    name="teacher",
+)
+
+# Create the student
+student = keras.Sequential(
+    [
+        keras.Input(shape=(28, 28, 1)),
+        layers.Conv2D(16, (3, 3), strides=(2, 2), padding="same"),
+        layers.LeakyReLU(alpha=0.2),
+        layers.MaxPooling2D(pool_size=(2, 2), strides=(1, 1), padding="same"),
+        layers.Conv2D(32, (3, 3), strides=(2, 2), padding="same"),
+        layers.Flatten(),
+        layers.Dense(10),
+    ],
+    name="student",
+)
+
+# Clone student for later comparison
+student_scratch = keras.models.clone_model(student)
+
+"""
+## Prepare dataset
+The dataset used for training the teacher and distilling the teacher is
+[MNIST](https://keras.io/api/datasets/mnist/), and the procedure is equivalent for other
+dataset, e.g. [CIFAR-10](https://keras.io/api/datasets/cifar10/), with a suitable choice
+of models. Both the student and teacher are trained on the training set and evaluated on
+the test set.
+"""
+
+# Prepare the train and test dataset.
+batch_size = 64
+(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
+
+# Normalize data
+x_train = x_train.astype("float32") / 255.0
+x_train = np.reshape(x_train, (-1, 28, 28, 1))
+
+x_test = x_test.astype("float32") / 255.0
+x_test = np.reshape(x_test, (-1, 28, 28, 1))
+
+# Construct datasets
+train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+train_dataset = train_dataset.shuffle(buffer_size=1024).batch(batch_size)
+
+test_dataset = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+test_dataset = test_dataset.shuffle(buffer_size=1024).batch(batch_size)
+
+"""
+## Train teacher
+In knowledge distillation we assume that the teacher is trained and fixed. Thus, we start
+by training the teacher model on the training set in the usual way.
+"""
+
+# Train teacher as usual
+teacher.compile(
+    optimizer=keras.optimizers.Adam(),
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    metrics=[keras.metrics.SparseCategoricalAccuracy()],
+)
+
+# Train and evaluate on subset of data to reduce computations.
+teacher.fit(train_dataset.take(10), epochs=5)
```

Comment:
Please train on the full dataset. It is a small model and a small dataset. It's only worth using a subset of the data if it takes 10+ min on GPU, which is not the case here.

That way you can also stick to using NumPy arrays for training.